tempodb/compactor.go

```diff
@@ -111,44 +113,48 @@ func (rw *readerWriter) doCompaction() {
 
 	level.Info(rw.logger).Log("msg", "starting compaction cycle", "tenantID", tenantID, "offset", rw.compactorTenantOffset)
 	for {
-		// Pick up to defaultMaxInputBlocks (4) blocks to compact into a single one
-		toBeCompacted, hashString := blockSelector.BlocksToCompact()
-		if len(toBeCompacted) == 0 {
-			measureOutstandingBlocks(tenantID, blockSelector, rw.compactorSharder.Owns)
-
-			level.Info(rw.logger).Log("msg", "compaction cycle complete. No more blocks to compact", "tenantID", tenantID)
-			break
-		}
-		if !rw.compactorSharder.Owns(hashString) {
-			// continue on this tenant until we find something we own
-			continue
-		}
-		level.Info(rw.logger).Log("msg", "Compacting hash", "hashString", hashString)
-		// Compact selected blocks into a larger one
-		err := rw.compact(toBeCompacted, tenantID)
-
-		if err == backend.ErrDoesNotExist {
-			level.Warn(rw.logger).Log("msg", "unable to find meta during compaction.  trying again on this block list", "err", err)
-		} else if err != nil {
-			level.Error(rw.logger).Log("msg", "error during compaction cycle", "err", err)
-			metricCompactionErrors.Inc()
-		}
-
-		// after a maintenance cycle bail out
-		if start.Add(rw.compactorCfg.MaxTimePerTenant).Before(time.Now()) {
-			measureOutstandingBlocks(tenantID, blockSelector, rw.compactorSharder.Owns)
-
-			level.Info(rw.logger).Log("msg", "compacted blocks for a maintenance cycle, bailing out", "tenantID", tenantID)
-			break
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			// Pick up to defaultMaxInputBlocks (4) blocks to compact into a single one
+			toBeCompacted, hashString := blockSelector.BlocksToCompact()
+			if len(toBeCompacted) == 0 {
+				measureOutstandingBlocks(tenantID, blockSelector, rw.compactorSharder.Owns)
+
+				level.Info(rw.logger).Log("msg", "compaction cycle complete. No more blocks to compact", "tenantID", tenantID)
+				return
+			}
+			if !rw.compactorSharder.Owns(hashString) {
+				// continue on this tenant until we find something we own
+				continue
+			}
+			level.Info(rw.logger).Log("msg", "Compacting hash", "hashString", hashString)
+			// Compact selected blocks into a larger one
+			err := rw.compact(ctx, toBeCompacted, tenantID)
+
+			if err == backend.ErrDoesNotExist {
+				level.Warn(rw.logger).Log("msg", "unable to find meta during compaction.  trying again on this block list", "err", err)
+			} else if err != nil {
+				level.Error(rw.logger).Log("msg", "error during compaction cycle", "err", err)
+				metricCompactionErrors.Inc()
+			}
+
+			// after a maintenance cycle bail out
+			if start.Add(rw.compactorCfg.MaxTimePerTenant).Before(time.Now()) {
+				measureOutstandingBlocks(tenantID, blockSelector, rw.compactorSharder.Owns)
+
+				level.Info(rw.logger).Log("msg", "compacted blocks for a maintenance cycle, bailing out", "tenantID", tenantID)
+				return
```

Comment:
No longer a break, but return.  Hard to see through the diff.  A couple other `break` is now `return`.